src/contrib/cluster/Akka.Cluster.Sharding/Delivery/Internal/ShardingConsumerControllerImpl.cs

```diff
@@ -115,7 +123,17 @@ private void Active()
         Receive<Terminated>(t => t.ActorRef.Equals(_consumer), _ =>
         {
             _log.Debug("Consumer terminated.");
-            Context.Stop(Self);
+            
+            // Short-circuit shutdown process, just shut down immediately if there's nothing to clean.
+            if (ProducerControllers.Count == 0 && ConsumerControllers.Count == 0)
+            {
+                _log.Debug("ShardingConsumerController terminated.");
+                Context.Stop(Self);
+            }
+            else
+            {
+                Become(ShuttingDown());
+            }
```

Comment:
Add a "shutting down" state to the actor, attempt cleanup before performing `Context.Stop(Self)`